### ISSUE TITLE
Dramatic slowdown

### ISSUE THREAD

**Issue by jbuckman** [alphatwirl/atpbar#8]
This progess bar enormously slows down my code for some reason. I'm attempting to use it to track gradient updates in Tensorflow. Running 2000 updates with a regular for-loop takes 8 seconds. Running 2000 updates with an atpbar takes 120 seconds. This is true even when I do:

```
for _ in atpbar.atpbar(range(10)):
    for _ in range(200):
        update()
```

so it's not related to the update rate of the bar.

In fact, even this code is slow:
```
for _ in atpbar.atpbar(range(10)): print("pass")
for _ in range(2000):
      update()
```

It seems the very act of initializing an atpbar is what slows things down. Any ideas what could be causing this behavior?

**Comment by TaiSakuma** [comment]
Thank you for letting me know. I will have a look.

**Comment by TaiSakuma** [comment]
I am trying to understand the slowness that you described.

The initialization of atpbar is immediate in my environment (macOS 10.15.1); it takes much less than 0.01 seconds.

Would you mind describing a bit more about your environment or how you run the code? For example, what does `update()` do in your code snippet? Is it possible to provide me with a minimum Python script file that reproduces the slowness?

**Comment by jbuckman** [comment]
Sorry for the delay -- I've put together a minimal example that replicates the behavior, and I have a partial understanding of what's going wrong. I think that atpbar is using too much CPU in the background, and it's slowing down my code.

Here's the code, I'll explain it below:
```
import tensorflow as tf
import numpy as np
import atpbar
import threading

import time

# how many times to run the experiment
N = 1000

# set up the queue
SHAPE = [32, 800]
x_placeholder = tf.placeholder(tf.float32, SHAPE)
tf_queue = tf.queue.FIFOQueue(capacity=100, dtypes=[tf.float32], shapes=[SHAPE])
enqueue_op = tf_queue.enqueue(x_placeholder)
enqueued_n = tf_queue.size()

# set up the computation graph
x = tf_queue.dequeue()
y = tf.random_uniform([800, 1000])
prod = tf.matmul(x,y)

# tf session, needed for executing operations in the computation graph
sess = tf.Session()

# function to fill up the queue
def fill():
  for i in range(N):
    sess.run(enqueue_op, feed_dict={x_placeholder: np.random.uniform(size=SHAPE)})

### FIRST TRIAL
filler = threading.Thread(target=fill, daemon=False) # launch the filler thread in the background
filler.start()
tt = time.time()
enqueued = []
for i in range(N):
  _, enqn = sess.run([prod, enqueued_n])
  enqueued.append(enqn > 0)
print("time in loop:", time.time() - tt)
print("% enqueued:", np.mean(enqueued) * 100)
filler.join()

### SECOND TRIAL
filler = threading.Thread(target=fill, daemon=False)
filler.start()
tt = time.time()
enqueued = []
for i in atpbar.atpbar(range(N)):
  _, enqn = sess.run([prod, enqueued_n])
  enqueued.append(enqn > 0)
print("time in atpbar loop:", time.time() - tt)
print("% enqueued:", np.mean(enqueued) * 100)
filler.join()

### THIRD TRIAL
filler = threading.Thread(target=fill, daemon=False) # launch the filler thread in the background
filler.start()
tt = time.time()
enqueued = []
for i in range(N):
  _, enqn = sess.run([prod, enqueued_n])
  enqueued.append(enqn > 0)
print("time in loop:", time.time() - tt)
print("% enqueued:", np.mean(enqueued) * 100)
filler.join()
```

This code uses a common pattern in TensorFlow: it uses a background thread to "enqueue" data and prepare it for processing, and then pulls data from the queue, and uses it in a computation. If the background thread is always faster at enqueueing data than the main thread is at using it, then the total runtime is just the time of the computation. However, if the background thread lags behind the main thread, then the code is bottlenecked by waiting for the data to be enqueued.

At the end of each trial, I print two things: the total time elapsed, and the percentage of the time that the data for the computation was immediately available to the main thread (i.e., the percentage of calculations that had zero blocking).

On my machine, the output of running this program looks like:

```
time in loop: 1.2820181846618652
% enqueued: 99.8
 100.00% :::::::::::::::::::::::::::::::::::::::: |     1000 /     1000 |:  range(0, 1000)
time in atpbar loop: 8.866333484649658
% enqueued: 13.700000000000001
time in loop: 9.71061635017395
% enqueued: 14.099999999999998
```

As you can see, the time to run has increased more than 10x when atpbar is used, and the culprit is the queuing: the regular for-loop almost never has to wait for more data, whereas the atpbar-loop finds itself blocked 85% of the time.

And, intriguingly, the *third* trial, which uses a regular for loop again, is also experiencing precisely the same slowdown! In fact, I've found that in general, if an atpbar.atpbar is created *anywhere* in the code, this same slowdown appears on every queue loop afterwards, whether a bar is used on the loop or not used or not.

I'm guessing atpbar itself uses threads in the background somehow, and that either these are just sucking up enough background CPU to slow things down, or they are somehow directly interfering with either the Python threading system or the TensorFlow queue logic.

Thanks for your help on this. Please let me know if you have any ideas!

**Comment by TaiSakuma** [comment]
Thank you for the detail. I will have a look when I can.

**Comment by TaiSakuma** [comment]
@jbuckman, thank you for the example code.

I released [v1.0.5](https://github.com/alphatwirl/atpbar/releases/tag/v1.0.5) to address the issue. Can you try it if you have a chance?

I tried running your code. The slowdown did occur on a Linux machine. (It didn't occur on Mac).

**[v1.0.4](https://github.com/alphatwirl/atpbar/releases/tag/v1.0.4)**

In v1.0.4, in fact, it would take much  longer to finish the loop once atpbar is enabled as you pointed out. In my environment, it took a lot more than 10 times longer.
```
time in loop: 2.0330605506896973
% enqueued: 85.8
 100.00% :::::::::::::::::::::::::::::::::::::::: |     1000 /     1000 |:  range(0, 1000)                                                                                                                 
time in atpbar loop: 107.39291834831238
% enqueued: 70.89999999999999
time in loop: 145.8734438419342
% enqueued: 52.900000000000006
```



**[v1.0.5](https://github.com/alphatwirl/atpbar/releases/tag/v1.0.5)**

In v1.0.5, it takes about the same time regardless of whether atpbar is enabled.
```
time in loop: 2.175316572189331
% enqueued: 47.099999999999994
 100.00% :::::::::::::::::::::::::::::::::::::::: |     1000 /     1000 |:  range(0, 1000)                                                                                                                 
time in atpbar loop: 2.21830415725708
% enqueued: 56.99999999999999
time in loop: 2.0305182933807373
% enqueued: 48.6
```

**Comment by jbuckman** [comment]
Yes, this has fixed the issue! Thanks for your help!
